src/main/java/org/openpnp/machine/reference/vision/ReferenceFiducialLocator.java

```diff
@@ -208,56 +264,152 @@ private static Location getFiducialLocation(BoardLocation boardLocation, Placeme
                     pkg.getId()));
         }
 
-        // Create the template
-        BufferedImage template = createTemplate(camera.getUnitsPerPixel(),
-                fid.getPart().getPackage().getFootprint());
-
         // Move to where we expect to find the fid
-        Location location =
+        Location expectedlocation =
                 Utils2D.calculateBoardPlacementLocation(boardLocation, fid.getLocation());
-        Logger.debug("Looking for {} at {}", fid.getId(), location);
-        MovableUtils.moveToLocationAtSafeZ(camera, location);
-
+        Logger.debug("Looking for {} at {}", fid.getId(), expectedlocation);
+        MovableUtils.moveToLocationAtSafeZ(camera, expectedlocation);
 
-        for (int i = 0; i < 3; i++) {
+        Location location = expectedlocation;
+        for (int i = 0; i < 3; i++) { // and the number of the counting shall be three
             // Wait for camera to settle
             Thread.sleep(camera.getSettleTimeMs());
             // Perform vision operation
-            location = getBestTemplateMatch(camera, template);
+            location = findFiducial(camera, part, boardLocation.getLocation().getRotation() + fid.getLocation().getRotation());
             if (location == null) {
                 Logger.debug("No matches found!");
-                return null;
+                continue;
             }
+            location = location.add(expectedlocation);
             Logger.debug("{} located at {}", fid.getId(), location);
             // Move to where we actually found the fid
-            camera.moveTo(location);
+            MovableUtils.moveToLocationAtSafeZ(camera, location);
+            break;
         }
 
         return location;
     }
 
-    private static Location getBestTemplateMatch(final Camera camera, BufferedImage template)
-            throws Exception {
-        VisionProvider visionProvider = camera.getVisionProvider();
+    // here's the big magic
+    // use default or custom pipeline for part to pinpoint fiducial location
+    private Location findFiducial(final Camera camera, Part part, double rotation) throws Exception {
+        BufferedImage template = null;
+        Object result = null;
+        
+        PartFiducialPipeline settings = getFiducialSettings(part);
+        CvPipeline pipeline = createPartPipeline(settings);
+        if (settings.isUseTemplateMatch()) {
+            // here we either need a footprint or a previously created template
+            Footprint fp = null;
+            if (rotation == settings.getTemplateRotation())
```

Comment:
Did I get it backwards?  It's supposed to not need re-creating the exact same template as required for the same part at the same rotation, trading a few CPU for memory.  It is an optimization, it can be removed.
However, I also allowed for not updating that template when viewing the fiducial pipeline, so it can be used as a model for the last part the camera viewed as a fiducial.  I think it works, and if there's no template yet created it makes one with angle 0.0D.  This was the spooky part I was joking about.  See if it works, but yeah, no trouble cutting it if it's unusual.